### ISSUE TITLE
Add a tool/script for this library

### ISSUE THREAD

**Issue by khushal-banks** [khushal-banks/log_bank#10]
- Hopefully to manage Makefile configuration
- Check if update is available
- Retrieve logs
- Search logs

**Comment by jgabaut** [comment]
> * Hopefully to manage Makefile configuration

I assume you're not interested in using `autotools` for `Makefile` configuration?

**Comment by khushal-banks** [comment]
> I assume you're not interested in using `autotools` for `Makefile` configuration?

I will have to check `autotools`.

Managing Makefiles would be just be an experiment.

I am planning to use either a bash script or a QT app using python (or c++) for doing this.
Something like a `log_bank UI` made for fun. :)

**Comment by jgabaut** [comment]
An UI just for configuring the build? I'd say go ahead and make a wholesale installer that also lets you pick target install path for the library dynamic/static objects and the headers.

I've been wanting to do a basic reusable one but haven't got around to doing so. I'm thinking of a basic install wizard, am sure you get what I mean. It includes configuration options in a page before the actual install, and builds it too.

**Comment by khushal-banks** [comment]
> I'm thinking of a basic install wizard
> configuration options in a page before the actual install
> An UI just for configuring the build?

Hmm .. you are right.. UI would need too much work and i am not very good at it. So, even if i create it .. it will not be portable.

Maybe, a single multi-purpose script would be enough.
I am used to scripting in bash but I will choose python (to learn new way of doing things)

**Comment by jgabaut** [comment]
> Maybe, a single multi-purpose script would be enough. I am used to scripting in bash but I will choose python (to learn new way of doing things)

Python is nice and would let you get the code out quickly, but you'd need to handle the python interpreter dependency and its quirks. I'd love to do one in C, since the only OS-dependent features it really can't do without would be things like creating directories, calling executables from paths (hence, some path-dependent logic). I don't have time right now but I will most likely try it sometime soon.

If you end up doing one in python, I'd love to try it out to use myself.

**Comment by khushal-banks** [comment]
> handle the python interpreter dependency and its quirks

Please add more. I don't think it would come in the way.

> OS-dependent features it really can't do without would be things like creating directories, calling executables from paths (hence, some path-dependent logic)

I highly doubt it. I will check more. But i know a bit of python it has APIs for everything you mentioned.

I genuinely think python won't pose any issue.

> I'd love to do one in C

I agree, C is my love too but we would end up creating a command line tool. 

Scripts however would be easily allow configuration using a json file or directly flags to control settings and path. Something like Open and Edit to configure they would also be platform independent.

**Comment by jgabaut** [comment]
> I agree, C is my love too but we would end up creating a command line tool. 

Not really, we are talking about a GUI tool. The envisioned code is the "common core" that would be included by would-be implementations (specific tools to install a specific project). So, not more of a command line tool than your typical bundled installer.exe with the install wizard GUI.

**Comment by khushal-banks** [comment]
> we are talking about a GUI tool. The envisioned code is the "common core"
> (specific tools to install a specific project)
> An UI just for configuring the build? I'd say go ahead and make a wholesale installer that also lets you pick target install path for the library dynamic/static objects and the headers.

Hmm .. my initial thought was that i would create either a tool/script which can list all possible [configuration parameters](https://github.com/khushal-banks/log_bank/blob/main/docs/COMPILATION.md) of this library. User will modify the value of these parameters directly from tool/script and the change would reflect in Makefile.

Then use can simply type make all and build the library.
I neglected installation directory because different user may use different project structure size (considering static linkage only). So, copying the files can be done by user manually because he know what he want.

And tool/script would validated entered data to show "invalid value of ... parameter" to help him understand [configuration parameter](https://github.com/khushal-banks/log_bank/blob/main/docs/COMPILATION.md) which he doesn't know.

And this tool would include other features:
1. Creating backup of log files to a tar ball to solve #11 
3. Filter wise searching logs to solve #10 (This issue is 10th issue)
3. Converting filter wise searched logs to an Excel file to solve #12 



> I don't have time right now but I will most likely try it sometime soon.

As much i understood your objective is a little different which is related to installation. right?

**Comment by jgabaut** [comment]
> Hmm .. my initial thought was that i would create either a tool/script which can list all possible configuration parameters [...]

This one is not really a concern since it can be derived easily from some file. 

> And tool/script would validated entered data to show "invalid value of ... parameter" to help him understand configuration parameter [...]

This brings me to ask the nature of each and every configuration parameter. For each option, does the makefile `-DEXAMPLE` definition:
- Just define a macro: `-DEXAMPLE_IS_DEFINED`
- Define a macro as an integer literal: `-DEXAMPLE=0`
-     as an string literal: `-DEXAMPLE="0"`
-     as a bool literal: `-DEXAMPLE=false` (can this even be done as a compiler argument? Never tried)
- Define a macro as another macro: `-DEXAMPLE_IS_DEFINED=MY_MACRO`
  - Which can lead to the other mentioned literals
- [...] Others

Knowing the needed ones might be useful, or not.

> And this tool would include other features:
> [...]
> As much i understood your objective is a little different which is related to installation. right?

I now understand you have clear requirements for your tool which are not what I was thinking of.

**Comment by khushal-banks** [comment]
> This brings me to ask the nature of each and every configuration parameter. For each option, does the makefile `-DEXAMPLE` definition:
> 
>     * Just define a macro: `-DEXAMPLE_IS_DEFINED`
> 
>     * Define a macro as an integer literal: `-DEXAMPLE=0`
> 
>     * (Copy above) as an string literal: `-DEXAMPLE="0"`
> 
>     * (Copy above) as a bool literal: `-DEXAMPLE=false` (can this even be done as a compiler argument? Never tried)
> 
>     * (Copy above) as another macro: `-DEXAMPLE_IS_DEFINED=MY_MACRO`
>       
>       * Which can lead to the other mentioned literals
> 
>     * [...] Others

Makefile and source code includes these checks.
This tool will validate the value before even user tried to compile the library.

For example, one may try to set some invalid readonly directory for log files or even a negative log-level.
User can definitely read [COMPILATION.md] but this tool will remove the hassle to refer docs
